{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scvelo as scv\n",
    "import loompy\n",
    "import igraph as ig\n",
    "from anndata import AnnData as ad\n",
    "import pandas as pd\n",
    "import re\n",
    "import scanpy as sc \n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np \n",
    "import scipy as sp\n",
    "\n",
    "adata = sc.read_h5ad(filename = \"/dkfz/groups/OE0540/users/ruehle/rnavelocity/velocyto/scvelo_output/use/dynamical/filteredpoordiff_cpmtop2000DE_pca3harmony_celltypes.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pseudotime mean aggregation at donor and run level, run separately per day of interest (doi)\n",
    "\n",
    "doi = \"day2\"\n",
    "\n",
    "day2 = adata[adata.obs['timepoint'] == doi, :]\n",
    "exp_day2 = day2.obs['run'].tolist()\n",
    "exp_day2 = list(dict.fromkeys(exp_day2))\n",
    "\n",
    "day0 = adata[adata.obs['timepoint'] == 'day0', :]\n",
    "day1 = adata[adata.obs['timepoint'] == 'day1', :]\n",
    "day2 = adata[adata.obs['timepoint'] == 'day2', :]\n",
    "day3 = adata[adata.obs['timepoint'] == 'day3', :]\n",
    "\n",
    "day0 = day0.obs[['donor', 'run', 'velocity_length']]\n",
    "day1 = day1.obs[['donor', 'run', 'velocity_length']]\n",
    "day2 = day2.obs[['donor', 'run', 'velocity_length']]\n",
    "day3 = day3.obs[['donor', 'run', 'velocity_length']]\n",
    "\n",
    "grouped_df = day2.groupby([\"run\", \"donor\"]).mean()\n",
    "day2 = grouped_df[grouped_df['velocity_length'].notna()]\n",
    "\n",
    "\n",
    "#obtain list of sample_names (for aggregated data)\n",
    "sample_ids = []\n",
    "\n",
    "for name in day2.index:\n",
    "    item = name[1] + \"-\" + name[0]\n",
    "    sample_ids.append(item)\n",
    "\n",
    "day2['feature_id'] = sample_ids\n",
    "day2 = day2.reset_index()\n",
    "day2 = day2[['feature_id', \"velocity_length\"]]\n",
    "\n",
    "day2.T.to_csv(r'/omics/groups/OE0540/internal/users/ruehle/rnavelocity/vqtls/data/velocyto/deterministic_length/top2000DE/input_day2/pheno_file_pseudotime_day2.txt', header=False, index=True, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>feature_id</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>start</td>\n",
       "      <td>end</td>\n",
       "      <td>ensembl_gene_id</td>\n",
       "      <td>feature_strand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>velocity_length</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>BLUB</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0           1      2    3                4               5\n",
       "0       feature_id  chromosome  start  end  ensembl_gene_id  feature_strand\n",
       "1  velocity_length           1      1    1             BLUB               +"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index = sample_ids\n",
    "exp_day2.insert(0, \"sample_id\")\n",
    "\n",
    "covariates = pd.DataFrame(index=index, columns=exp_day2)\n",
    "covariates['sample_id'] = sample_ids\n",
    "covariates = covariates.fillna(0) # with 0s\n",
    "covariates\n",
    "\n",
    "#fill covariance matrix \n",
    "#if substring in fullstring:\n",
    "\n",
    "for index, row in covariates.iterrows():\n",
    "    for column in covariates:\n",
    "        if column in index:\n",
    "            covariates.loc[index, column] = 1\n",
    "covariates\n",
    "\n",
    "covariates.to_csv(r'/omics/groups/OE0540/internal/users/ruehle/rnavelocity/vqtls/data/velocyto/deterministic_length/top2000DE/input_day2/covariance_pseudotime_day2.txt', header=True, index=False, sep='\\t', mode='a')\n",
    "\n",
    "feature_anno_file = pd.DataFrame()\n",
    "#Feature annotation file\n",
    "data = [['feature_id', 'chromosome', 'start', 'end', 'ensembl_gene_id', 'feature_strand'], ['velocity_length', 1, 1, 1, 'BLUB', '+']]\n",
    "feature_anno_file = pd.DataFrame(data)\n",
    "#feature_anno_file.head()\n",
    "feature_anno_file.to_csv(r'/omics/groups/OE0540/internal/users/ruehle/rnavelocity/vqtls/data/velocyto/deterministic_length/feature_anno_file_pc_pseudotime.txt', header=False, index=False, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#link file alias sample mapping file\n",
    "#maps genotype names to sample names \n",
    "\n",
    "donormeta = pd.read_csv('/icgc/dkfzlsdf/analysis/B260/projects/HipSci/openAccess/endoderm_differentation/metadata/HipSci_donor_annotation.tsv', sep = ';', header = 0)\n",
    "\n",
    "subdf = pd.DataFrame()\n",
    "final = pd.DataFrame()\n",
    "subdf['sample_id'] = day2.T.loc[\"feature_id\"]\n",
    "\n",
    "subdf['sample_map'] = \"zero\"\n",
    "subdf['donor'] = \"donor\"\n",
    "\n",
    "#remove run donor thing and put it into column donor of subdf \n",
    "for index, row in subdf.iterrows():\n",
    "    donor_o = str(row['sample_id'])\n",
    "    donor_only = donor_o[:-6]\n",
    "    row['donor'] = donor_only\n",
    "\n",
    "for sample_index, sample_row in subdf.iterrows():\n",
    "    sample = sample_row['sample_id']\n",
    "    for genomic_index, genomic_row in donormeta.iterrows():\n",
    "        if genomic_row['DonorLine'] in sample:\n",
    "            sample_row['sample_map'] = genomic_row[\"Sample2\"]\n",
    "        else:\n",
    "            None\n",
    "\n",
    "final = subdf[['sample_map', 'sample_id']]\n",
    "#final.head()\n",
    "\n",
    "final.to_csv(r'/omics/groups/OE0540/internal/users/ruehle/rnavelocity/vqtls/data/velocyto/deterministic_length/top2000DE/input_day2/sample_mapping_file_pseudotime_day2.txt', header=False, index=False, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SCVELO",
   "language": "python",
   "name": "scvelo"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
